enhancements/ingress/haproxy-log-format.md

```diff
@@ -0,0 +1,272 @@
+collected in separate logs.
+
+## Proposal
+
+The `AccessLogging` type allows specifying a destination using the `Destination`
+field, which has type `LoggingDestination`, and the log format using the
+`HttpLogFormat` field, which has type `string`:
+
+```go
+// AccessLogging describes how client requests should be logged.
+type AccessLogging struct {
+        // destination is where access logs go.
+        //
+        // +kubebuilder:validation:Required
+        // +required
+        Destination LoggingDestination `json:"destination"`
+
+        // httpLogFormat specifies the format of the log message for an HTTP
+        // request.
+        //
+        // If this field is empty, log messages use the implementation's default
+        // HTTP log format.  For HAProxy's default HTTP log format, see the
+        // HAProxy documentation:
+        // http://docs.haproxy.org/2.8/configuration.html#8.2.3
+        //
+        // +optional
+        HttpLogFormat string `json:"httpLogFormat,omitempty"`
+
+        // httpsLogFormat specifies the format of the log messages for
+        // an HTTPS request
+        //
+        // If this field is empty, log messages use the implementation's default
+        // HTTPS log format.  For HAProxy's default HTTPS log format, see the
+        // HAProxy documentation:
+        // http://docs.haproxy.org/2.8/configuration.html#8.2.4
+        //
+        // +optional
+        HttpsLogFormat string `json:"httpsLogFormat,omitempty"`
+
+        // tcpLogFormat specifies the format of the log messages for a
+        // TCP request.
+        //
+        // If this field is empty, log messages use the implementation's default
+        // TCP log format.  For HAProxy's default TCP log format, see the
+        // HAProxy documentation:
+        // http://docs.haproxy.org/2.8/configuration.html#8.2.2
+        //
+        // +optional
+        TcpLogFormat string `json:"tcpLogFormat,omitempty"`
+}
+```
+
+The user may specify `spec.logging.access.httpLogFormat` to customize
+the HTTP log format. The user may also specify
+`spec.logging.access.httpsLogFormat` to customize the HTTPS log
+format. The user may also specify `spec.logging.access.tcpLogFormat`
+to customize TCP log format. For example, the following is the
+definition for an IngressController that customizes each of the log
+formats (HTTP, HTTPS, and TCP):
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  replicas: 2
+  endpointPublishingStrategy:
+    type: NodePortService
+  logging:
+    access:
+      httpLogFormat: "%ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC ..."
+      httpsLogFormat: "%ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC ..."
+      tcpLogFormat: "%ci:%cp [%t] %ft %b/%s %Tw/%Tc/%Tt %B %ts ..."
+```
+
+Each of the log format fields contains a free-form format string, so
+none of these are validated.
```

Comment:
Can we consider doing some form of validation, or at least adding unit tests that show malformatted log formats don't break the router?